From: Jakub Zelenka <bukka@php.net>
Date: Fri, 29 Mar 2024 15:27:59 +0000
Subject: Fix bug GHSA-q6x7-frmf-grcw: password_verify can erroneously return
 true

Disallow null character in bcrypt password

(cherry picked from commit 0ba5229a3f7572846e91c8f5382e87785f543826)
(cherry picked from commit 81794c73068d9a44bf109bbcc9793e7b56a1c051)
(cherry picked from commit 4a7ceb9d6427f8d368f1a8739267b1f8310ec201)
(cherry picked from commit 747100905eceffb1f67096b437001e42900eb6bb)
(cherry picked from commit d22d9ebb29dce86edd622205dd1196a2796c08c7)
(cherry picked from commit cd9a376c28c6f4ce83aab53ec069234fe1d2a819)
(cherry picked from commit 459b4ac6a8d9bec32110b68ac194d71ec2b72182)
(cherry picked from commit 275555907e5c56b63ca2ae3259009e5537f970e5)
---
 ext/standard/password.c                                 | 7 ++++++-
 ext/standard/tests/password/password_bcrypt_errors.phpt | 4 ++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/ext/standard/password.c b/ext/standard/password.c
index 5a8edad..757d086 100644
--- a/ext/standard/password.c
+++ b/ext/standard/password.c
@@ -334,7 +334,12 @@ PHP_FUNCTION(password_hash)
 					cost = Z_LVAL_PP(option_buffer);
 				}
 			}
-	
+
+			if (memchr(password, '\0', password_len)) {
+				php_error_docref(NULL TSRMLS_CC, E_WARNING, "Bcrypt password must not contain null character");
+				RETURN_NULL();
+			}
+
 			if (cost < 4 || cost > 31) {
 				php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid bcrypt cost parameter specified: %ld", cost);
 				RETURN_NULL();
diff --git a/ext/standard/tests/password/password_bcrypt_errors.phpt b/ext/standard/tests/password/password_bcrypt_errors.phpt
index 2548c9a..617f468 100644
--- a/ext/standard/tests/password/password_bcrypt_errors.phpt
+++ b/ext/standard/tests/password/password_bcrypt_errors.phpt
@@ -16,6 +16,8 @@ var_dump(password_hash("foo", PASSWORD_BCRYPT, array("salt" => 123)));
 
 var_dump(password_hash("foo", PASSWORD_BCRYPT, array("cost" => "foo")));
 
+var_dump(password_hash("null\0password", PASSWORD_BCRYPT));
+
 ?>
 --EXPECTF--
 Warning: password_hash(): Invalid bcrypt cost parameter specified: 3 in %s on line %d
@@ -36,4 +38,6 @@ NULL
 Warning: password_hash(): Invalid bcrypt cost parameter specified: 0 in %s on line %d
 NULL
 
+Warning: password_hash(): Bcrypt password must not contain null character in %s on line %d
+NULL
 
